pkg/k8s/custom_source.go

```diff
@@ -0,0 +1,181 @@
+package k8s
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+const (
+	// defaultBufferSize is the default number of event notifications that can be buffered.
+	defaultBufferSize = 1024
+)
+
+// Source is a source of events
+type Source interface {
+	// Start is internal and should be called only by the Controller to register an EventHandler with the Informer
+	// to enqueue reconcile.Requests.
+	Start(handler.EventHandler, workqueue.RateLimitingInterface, ...predicate.Predicate) error
+}
+
+var _ Source = &NotificationChannel{}
+
+// EventType identifies type of PodEvent (CREATE/UPDATE/DELETE)
+type EventType int
+
+const (
+	CREATE EventType = 1
+	DELETE EventType = 2
+	UPDATE EventType = 3
+)
+
+// PodEvent is a wrapper for Create/Update/Delete Events
+type PodEvent struct {
+	EventType
+
+	// Meta is the ObjectMeta of the Kubernetes Type that was created
+	Meta metav1.Object
+
+	// Object is the object from the event
+	Object runtime.Object
+
+	// MetaOld is the ObjectMeta of the Kubernetes Type that was updated (before the update)
+	MetaOld metav1.Object
```

Comment:
Meta      -> for the create/update's new oldObj
Object   -> for the create/update's new oldObj

OldMeta     -> for the delete or update's old oldObj
OldObject  -> for the delete or update's old oldObj



for example k8s uses below.

```
         // Object is the object from the incoming request.
	// +optional
	Object runtime.RawExtension `json:"object,omitempty" protobuf:"bytes,9,opt,name=object"`
	// OldObject is the existing object. Only populated for DELETE and UPDATE requests.
	// +optional
	OldObject runtime.RawExtension `json:"oldObject,omitempty" protobuf:"bytes,10,opt,name=oldObject"`
```